prow/pluginhelp/hook/hook.go

```diff
@@ -0,0 +1,366 @@
+		b, err := json.Marshal(enabledRepos)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling enabled repos: %q, err: %v", enabledRepos, err)
+		}
+
+		// Don't retry because user is waiting for response to their browser.
+		// If there is an error the user can refresh to get the plugin info that failed to load.
+		urlString := url.String()
+		resp, err := http.Post(urlString, "application/json", bytes.NewReader(b))
+		if err != nil {
+			return nil, fmt.Errorf("error posting to %s err: %v", urlString, err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode < 200 || resp.StatusCode > 299 {
+			return nil, fmt.Errorf("post to %s failed with status %d: %s", urlString, resp.StatusCode, resp.Status)
+		}
+		var help pluginhelp.PluginHelp
+		if err := json.NewDecoder(resp.Body).Decode(&help); err != nil {
+			return nil, fmt.Errorf("failed to decode json response from %s err: %v", urlString, err)
+		}
+		return &help, nil
+	}
+}
+
+// reversePluginMaps inverts the Configuration.Plugins and Configuration.ExternalPlugins maps and
+// expands any org strings to org/repo strings.
+// The returned values map plugin names to the set of org/repo strings they are enabled on.
+func reversePluginMaps(config *plugins.Configuration, orgToRepos map[string]sets.String) (normal, external map[string][]string) {
+	normal = map[string][]string{}
+	for repo, plugins := range config.Plugins {
+		var repos []string
+		if flattened, ok := orgToRepos[repo]; ok {
+			repos = flattened.List()
+		} else {
+			repos = []string{repo}
+		}
+		for _, plugin := range plugins {
+			normal[plugin] = append(normal[plugin], repos...)
+		}
+	}
+	external = map[string][]string{}
+	for repo, plugins := range config.ExternalPlugins {
+		var repos []string
+		if flattened, ok := orgToRepos[repo]; ok {
+			repos = flattened.List()
+		} else {
+			repos = []string{repo}
+		}
+		for _, plugin := range plugins {
+			external[plugin.Name] = append(external[plugin.Name], repos...)
+		}
+	}
+	return
+}
+
+// orgAgent provides a cached mapping of orgs to the repos that are in that org.
+// Caching is necessary to prevent excessive github API token usage.
+type orgAgent struct {
+	log *logrus.Entry
+	ghc githubClient
+
+	syncPeriod time.Duration
+
+	lock       sync.Mutex
+	nextSync   time.Time
+	orgs       sets.String
+	orgToRepos map[string]sets.String
+}
+
+func newOrgAgent(log *logrus.Entry, ghc githubClient, syncPeriod time.Duration) *orgAgent {
+	return &orgAgent{
+		log:        log,
+		ghc:        ghc,
+		syncPeriod: syncPeriod,
+	}
+}
+
+func (oa *orgAgent) orgToReposMap(config *plugins.Configuration) map[string]sets.String {
+	oa.lock.Lock()
+	defer oa.lock.Unlock()
+	// Only need to sync if either:
+	// - the sync period has passed (we sync periodically to pick up new repos in known orgs)
+	//  or
+	// - new org(s) have been added in the config
+	if time.Now().After(oa.nextSync) || orgsInConfig(config).Difference(oa.orgs).Len() > 0 {
+		oa.sync(config)
```

Comment:
Log every time we sync. If this ends up being noisy we can make it Debug.

Reply:
Done.